sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,276 @@
+all non-`SYCL_EXTERNAL` functions are internalized to avoid multiple definition
+errors during runtime linking.
+Device images with `SYCL_EXTERNAL` functions will also have a list of names
+of exported functions.
+
+**NOTE**: If device code split is enabled, it seems reasonable to perform
+exports arrangement before device code split procedure.
+
+In orger to collect information about imported symbols `sycl-post-link` looks
+through LLVM IR and for each declared but not defined symbol records its name,
+except the following cases:
+- Declarations with `__` prefix in demangled name are not recorded as imported
+  functions
+  - Declarations with `__spirv_*` prefix should not be recorded as dependencies
+  since they represent SPIR-V operations and will be transformed to SPIR-V
+  instructions during LLVM->SPIR-V translation.
+- Based on some attributes which could be defined later
+  - This is needed to have possibility to call device-specific builtins not
+    starting with `__` by forward-declaring them in DPC++ code
+
+**NOTE**: If device code split is enabled, imports collection is performed after
+split and it is performed on splitted images.
+
+All collected information is attached to a device image via properties
+mechanism.
+
+Each device image is supplied with an array of property sets:
+```
+struct pi_device_binary_struct {
+...
+  // Array of property sets
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+
+```
+Each property set is represent by the following struct:
+```
+// Named array of properties.
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+```
+It contains name of property set and array of properties. Each property is
+represented by the following struct:
+```
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+List of imported symbols is represented as a single property set with name
+`ImportedSymbols` recorded in the `Name` field of property set.
+Each property in this set holds name of the particular imported symbol recorded
+in the `Name` field of the property.
+List of exported symbols is represented in the same way, except the
+corresponding set has the name `ExportedSymbols`.
+
+### DPC++ runtime changes
+
+DPC++ RT performs *device images collection* task by grouping all device
+images required to execute a kernel based on the list of exports/imports and
+links them together using PI API.
+
+Given that all exports will be arranged to a separate device images without
+kernels it is reasonable to store device images with exports in a separate data
+structure.
+
```

Comment:
Please specify resolution process in more details. Please address a situation when there are multiple host shared objects in addition to executables, with multiple device binary images in each. The same name can be exported from multiple images. How runtime finds images to satisfy a dependence? Uses the first found? A lot depends on module-id here i guess, so it should be specified in full detail (I also requested this above). In practice there can be 1000s of modules (or even more) when splitting is engaged, so dependence resolution performance will likely matter.

Reply:
> How runtime finds images to satisfy a dependence? Uses the first found?
It uses the first found.

> The same name can be exported from multiple images.

If these images export the same function under this name, then it is not important which device image will be taken (since the symbol is same). If these same named functions are different, then ODR violation takes place. This is undefined behavior. We don't guarantee that it will work as user expects.